app/code/Magento/Inventory/Observer/SourceItemsProcessor.php

```diff
@@ -0,0 +1,158 @@
+        SourceItemInterfaceFactory $sourceItemFactory,
+        DataObjectHelper $dataObjectHelper,
+        SourceItemSaveInterface $sourceItemSave
+    ) {
+        $this->searchCriteriaBuilderFactory = $searchCriteriaBuilderFactory;
+        $this->sourceItemRepository = $sourceItemRepository;
+        $this->sourceItemFactory = $sourceItemFactory;
+        $this->dataObjectHelper = $dataObjectHelper;
+        $this->sourceItemSave = $sourceItemSave;
+    }
+
+    /**
+     * @param string $sku
+     * @param array $sourceItemsData
+     * @return void
+     * @throws InputException
+     */
+    public function process($sku, array $sourceItemsData)
+    {
+        $sourceItemsForDelete = $this->getCurrentSourceItemsMap($sku);
+        $sourceItemsForSave = [];
+
+        foreach ($sourceItemsData as $sourceItemData) {
+            $this->validateSourceItemData($sourceItemData);
+
+            $sourceId = $sourceItemData[SourceItemInterface::SOURCE_ID];
+            if (isset($sourceItemsForDelete[$sourceId])) {
+                $sourceItem = $sourceItemsForDelete[$sourceId];
+            } else {
+                /** @var SourceItemInterface $relation */
+                $sourceItem = $this->sourceItemFactory->create();
+            }
+
+            $sourceItemData[SourceItemInterface::SKU] = $sku;
+            $this->dataObjectHelper->populateWithArray($sourceItem, $sourceItemData, SourceItemInterface::class);
+
+            $sourceItemsForSave[] = $sourceItem;
+            unset($sourceItemsForDelete[$sourceId]);
+        }
+        $this->saveSourceItems($sourceItemsForSave);
+        $this->deleteSourceItems($sourceItemsForDelete);
+    }
+
+    /**
+     * Key is source id, value is Source Item
+     *
+     * @param string $sku
+     * @return SourceItemInterface[]
+     */
+    private function getCurrentSourceItemsMap($sku)
+    {
+        /** @var SearchCriteriaBuilder $searchCriteriaBuilder */
+        $searchCriteriaBuilder = $this->searchCriteriaBuilderFactory->create();
+        $searchCriteria = $searchCriteriaBuilder
+            ->addFilter(ProductInterface::SKU, $sku)
+            ->create();
+        $sourceItems = $this->sourceItemRepository->getList($searchCriteria)->getItems();
+
+        $sourceItemMap = [];
+        if ($sourceItems) {
+            foreach ($sourceItems as $sourceItem) {
+                $sourceItemMap[$sourceItem->getSourceId()] = $sourceItem;
+            }
+        }
+        return $sourceItemMap;
+    }
+
+    /**
+     * @param array $sourceItemData
+     * @return void
+     * @throws InputException
+     */
+    private function validateSourceItemData(array $sourceItemData)
+    {
+        if (!isset($sourceItemData[SourceItemInterface::SOURCE_ID])) {
+            throw new InputException(__('Wrong Product to Source relation parameters.'));
+        }
+    }
+
+    /**
+     * @param array $sourceItems
+     * @return void
+     */
+    private function saveSourceItems(array $sourceItems)
```

Comment:
why do we need to extract separate method for such simple and easy to read operation ?